docs/lazy-adr/adr-003-application-data-retrieval.md

```diff
@@ -0,0 +1,137 @@
+# ADR 003: Retrieving Application messages
+
+## Changelog
+
+- 2021-04-25: initial draft
+
+## Context
+
+The academic Lazyledger [paper](https://arxiv.org/abs/1905.09274) describes the motivation and context for this API.
+The main motivation can be quoted from section 3.3 of that paper:
+
+> (Property1) **Application message retrieval partitioning.** Client nodes must be able to download all of the messages relevant to the applications they use [...], without needing to downloading any messages for other applications.
+
+> (Property2) **Application message retrieval completeness.** When client nodes download messages relevant to the applications they use [...], they must be able to verify that the messages they received are the complete set of messages relevant to their applications, for specific
+blocks, and that there are no omitted messages.
+
+
+
+The main data structure that enables above properties is called a Namespaced Merkle Tree (NMT), an ordered binary Merkle tree where:
+1. each node in the tree includes the range of namespaces of the messages in all descendants of each node
+2. leaves in the tree are ordered by the  namespace identifiers of the leaf messages
+
+A more formal description can be found the [specification](https://github.com/lazyledger/lazyledger-specs/blob/de5f4f74f56922e9fa735ef79d9e6e6492a2bad1/specs/data_structures.md#namespace-merkle-tree).
+An implementation can be found in [this repository](https://github.com/lazyledger/nmt).
+
+This ADR basically describes version of the [`GetWithProof`](https://github.com/lazyledger/nmt/blob/ddcc72040149c115f83b2199eafabf3127ae12ac/nmt.go#L193-L196) of the NMT that leverages the fact that IPFS uses content addressing and that we have implemented an [IPLD plugin](https://github.com/lazyledger/lazyledger-core/tree/37502aac69d755c189df37642b87327772f4ac2a/p2p/ipld) for an NMT.
+
+**Note**: The APIs defined here will be particularly relevant for Optimistic Rollup (full) nodes that want to download their Rollup's data (see [lazyledger/optimint#48](https://github.com/lazyledger/optimint/issues/48)).
```

Comment:
Good point, will add that! Now that you've mentioned it, I guess we could also use a modified version of tendermint's internal gossiping mechanism for that: currently, the gossip routine pushes out random chunks of the whole (regular) block data (aka Parts) but instead, it could send out only the state relevant portions instead (as namespaced shares). 

I feel like this could have several benefits: this gossiping is potentially faster than requesting the data from ipfs, the code is already there (but would need to be modified), we would remove some pressure from the proposer more quickly as we decrease the time where it is the only node on the network having all block data; it would buy some time for doing the DA sampling in parallel or pipelined shortly after receiving a few shares from the gossip routine.

Reply:
Added use-case: 916b289

I think the idea of using the gossiping mechanism should be discussed in an ADR / issue. Will leave the comment unresolved s.t. I can quickly find it when going through the open comments here.

Reply:
> the gossip routine pushes out random chunks of the whole (regular) block data (aka Parts) but instead, it could send out only the state relevant portions instead (as namespaced shares).

Hmm, that's an interesting idea and for its benefits, it can be better than what I was doing. I was recently thinking about/partially implementing that, but instead of parts, I am sending DAHeaders for nodes to catch up.

But I am concerned this is not a long-term solution and may require changes that soon will be removed, as IPFS is our main data source/sharing mechanism. Also, this is quite a sharp turn and should be documented as an issue somewhere first

Reply:
> Also, this is quite a sharp turn and should be documented as an issue somewhere first

Yes, but previously we were not optimizing for the use-case of light validator nodes that need to download all state relevant Txs. I guess with the deferred execution model that doesn't matter much as validators can start downloading these while the previous block is reaching consensus (it's the previous block's Tx the app hash is computed for) but with immediate execution, this will become a time-sensitive (and kinda consensus critical) issue. I think a simple hybrid gossiping approach that pushes out some data that every validator will need anyway paired with IPFS to get all other data on demand (e.g. for DAS or for downloading the whole block) could be a better approach. I don't think we have decided on any particular approach yet anyways.

Reply:
Just noting that this is not a decision either. I'm proposing to _investigate_ this further: e.g. by looking into how easy it would be to change the current gossiping routine to only gossip reserved Txs only. If it turns out not to be easy we could decide to go down the other route.